src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/tools/jcore/ByteCodeRewriter.java

```diff
@@ -152,22 +116,29 @@ public void rewrite() {
                 case Bytecodes._getstatic:
                 case Bytecodes._putstatic:
                 case Bytecodes._getfield:
-                case Bytecodes._putfield:
+                case Bytecodes._putfield: {
+                    int fieldIndex = method.getNativeShortArg(bci + 1);
+                    cpoolIndex = (short) cpCache.getFieldEntryAt(fieldIndex).getConstantPoolIndex();
+                    writeShort(code, bci + 1, cpoolIndex);
+                    break;
+                }
                 case Bytecodes._invokevirtual:
                 case Bytecodes._invokespecial:
                 case Bytecodes._invokestatic:
                 case Bytecodes._invokeinterface: {
-                    cpoolIndex = getConstantPoolIndex(hotspotcode, bci + 1);
+                    int cpci = method.getNativeShortArg(bci + 1);
+                    cpoolIndex = (short) cpCache.getEntryAt(cpci).getConstantPoolIndex();
                     writeShort(code, bci + 1, cpoolIndex);
                     break;
                 }
 
-                case Bytecodes._invokedynamic:
-                    cpoolIndex = getConstantPoolIndex(hotspotcode, bci + 1);
+                case Bytecodes._invokedynamic: {
+                    int cpci = method.getNativeIntArg(bci + 1);
+                    cpoolIndex = (short) cpCache.getIndyEntryAt(~cpci).getConstantPoolIndex();
```

Comment:
Is this really suppose to be `~cpci` and not just `cpci`?

Reply:
That's right. See https://github.com/openjdk/jdk/blob/ecb5e8a03f67c92d7956201de1fa7d07cc6af9cb/src/hotspot/share/prims/jvmtiClassFileReconstituter.cpp#L1053 and https://github.com/openjdk/jdk/blob/ecb5e8a03f67c92d7956201de1fa7d07cc6af9cb/src/hotspot/share/oops/constantPool.hpp#L256

I tend to use `JvmtiClassFileReconstituter` as the reference for the code involved in dumping the classfile, and that makes things easier to follow and implement.